### PR TITLE
AV-196565 commit for L7Rule crd code

### PR DESCRIPTION
**Pull request by arihantg**
Manually Tested below scenarios in Dedicated and Large EVH mode

1. L7Rule CRD name Add/Update/delete in hostrule
2. L7Rule CRD Create/deletion/Update
3. invalid to valid transition of L7 CRd

### CODE REVIEW

**Comment by arihantg** [comment]
build ako

**Comment by arihantg** [comment]
build ako

File: internal/k8s/validator.go [hunk 1]
```diff
@@ -717,6 +727,37 @@ func (l *leader) ValidateL4RuleObj(key string, l4Rule *akov1alpha2.L4Rule) error
...
+		refData[*l7RuleSpec.TrafficCloneProfileRef] = "TrafficCloneProfile"
+	}
+	if err := checkRefsOnController(key, refData); err != nil {
+		status.UpdateL7RuleStatus(key, l7Rule, status.UpdateCRDStatusOptions{
+			Status: lib.StatusRejected,
+			Error:  err.Error(),
+		})
+		return err
+	}
+	// No need to update status of l4rule object as accepted since it was accepted before.
```
Comment (pkoshtavmware) [inline internal/k8s/validator.go:753]: might want to change l4rule to l7rule here

File: internal/nodes/avi_model_nodes_utils.go [hunk 2]
```diff
@@ -151,6 +202,30 @@ func (o *AviVsNodeCommonFields) ConvertToRef() {
 	}
 }
 
+func (o *AviVsNodeGeneratedFields) ConvertL7RuleFieldsToNil() {
+	if o != nil {
+		o.AllowInvalidClientCert = nil
+		o.BotPolicyRef =nil
+		o.CloseClientConnOnConfigUpdate =nil
```
Comment (pkoshtavmware) [inline internal/nodes/avi_model_nodes_utils.go:209]: Is it ok to set boolean fields to nil as well ?

File: pkg/client/v1alpha2/clientset/versioned/typed/ako/v1alpha2/fake/fake_l4rule.go [hunk 3]
```diff
@@ -36,9 +35,9 @@ type FakeL4Rules struct {
 	ns   string
 }
 
-var l4rulesResource = schema.GroupVersionResource{Group: "ako.vmware.com", Version: "v1alpha2", Resource: "l4rules"}
+var l4rulesResource = v1alpha2.SchemeGroupVersion.WithResource("l4rules")
```
Comment (pkoshtavmware) [inline pkg/client/v1alpha2/clientset/versioned/typed/ako/v1alpha2/fake/fake_l4rule.go:38]: Not sure why this is changed, if changes are related to l7Rules

File: pkg/client/v1alpha2/clientset/versioned/typed/ako/v1alpha2/fake/fake_ssorule.go [hunk 4]
```diff
@@ -36,9 +35,9 @@ type FakeSSORules struct {
 	ns   string
 }
 
-var ssorulesResource = schema.GroupVersionResource{Group: "ako.vmware.com", Version: "v1alpha2", Resource: "ssorules"}
+var ssorulesResource = v1alpha2.SchemeGroupVersion.WithResource("ssorules")
```
Comment (pkoshtavmware) [inline pkg/client/v1alpha2/clientset/versioned/typed/ako/v1alpha2/fake/fake_ssorule.go:38]: Same as above

**Review by pkoshtavmware — COMMENTED** [review]
Overall changes looks good to me except a few comments.
Also could you please attach information on how testing is done

**Comment by arihantg** [comment]
build ako

**Comment by arihantg** [comment]
build ako

File: go.sum [hunk 5]
```diff
@@ -32,6 +32,7 @@ cloud.google.com/go/storage v1.8.0/go.mod h1:Wv1Oy7z6Yz3DshWRJFhqM/UCfaWIRTdp0RX
 cloud.google.com/go/storage v1.10.0/go.mod h1:FLPqc6j+Ki4BU591ie1oL6qBQGu2Bl/tZ9ullr3+Kg0=
 dmitri.shuralyov.com/gpu/mtl v0.0.0-20190408044501-666a987793e9/go.mod h1:H6x//7gZCb22OMCxBHrMx7a5I7Hp++hsVxbQ4BYO7hU=
 github.com/Azure/go-ansiterm v0.0.0-20170929234023-d6e3b3328b78/go.mod h1:LmzpDX56iTiv29bbRTIsUNlaFfuhWRQBWjQdVyAevI8=
+github.com/Azure/go-ansiterm v0.0.0-20210617225240-d185dfc1b5a1/go.mod h1:xomTg63KZ2rFqZQzSB4Vz2SUXa1BpHTVz9L5PTmPC4E=
```
[the remark this answers is not in the document]
Comment (arihantg) [reply go.sum]: updated the file after executing go mod tidy

File: internal/k8s/validator.go (hunk 1 above)
Comment (arihantg) [reply internal/k8s/validator.go]: changed

File: internal/nodes/avi_model_nodes_utils.go (hunk 2 above)
Comment (arihantg) [reply internal/nodes/avi_model_nodes_utils.go:209]: yes. In UT we test them for nil value as well.

File: pkg/client/v1alpha2/clientset/versioned/typed/ako/v1alpha2/fake/fake_l4rule.go (hunk 3 above)
Comment (arihantg) [reply pkg/client/v1alpha2/clientset/versioned/typed/ako/v1alpha2/fake/fake_l4rule.go]: Autogeneration is done for  all crds in v1aplha2 version. So this part in L4Rule and SS0Rule is also changed.

File: pkg/client/v1alpha2/clientset/versioned/typed/ako/v1alpha2/fake/fake_ssorule.go (hunk 4 above)
Comment (arihantg) [reply pkg/client/v1alpha2/clientset/versioned/typed/ako/v1alpha2/fake/fake_ssorule.go]: Autogeneration is done for all crds in v1aplha2 version. So this part in L4Rule and SS0Rule is also changed.

File: internal/k8s/validator.go [hunk 6]
```diff
@@ -40,6 +40,7 @@ type Validator interface {
 	ValidateServiceImportObj(key string, serviceImport *akov1alpha1.ServiceImport) error
 	ValidateSSORuleObj(key string, ssoRule *akov1alpha2.SSORule) error
 	ValidateL4RuleObj(key string, l4Rule *akov1alpha2.L4Rule) error
+	ValidateL7RuleObj(key string, l4Rule *akov1alpha2.L7Rule) error
```
[the remark this answers is not in the document]
Comment (arihantg) [reply internal/k8s/validator.go]: changed

File: internal/k8s/validator.go [hunk 7]
```diff
@@ -799,3 +840,8 @@ func (l *follower) ValidateL4RuleObj(key string, l4Rule *akov1alpha2.L4Rule) err
 	utils.AviLog.Debugf("key: %s, AKO is not a leader, not validating L4Rule object", key)
 	return nil
 }
+
+func (f *follower) ValidateL7RuleObj(key string, l4Rule *akov1alpha2.L7Rule) error {
```
[the remark this answers is not in the document]
Comment (arihantg) [reply internal/k8s/validator.go]: changed

File: internal/nodes/avi_model_nodes_utils.go [hunk 8]
```diff
@@ -28,28 +28,61 @@ import (
...
-	SamlSpConfig             *v1alpha2.SAMLSPConfig
-	SecurityPolicyRef        *string
+	AllowInvalidClientCert        *bool
+	BotPolicyRef                  *string
+	CloseClientConnOnConfigUpdate *bool
+	Fqdn                          *string
+	HostNameXlate                 *string
+	IgnPoolNetReach               *bool
+	LoadBalancerIP                *string
+	MinPoolsUp                    *int32
```
[the remark this answers is not in the document]
Comment (arihantg) [reply internal/nodes/avi_model_nodes_utils.go]: updated to *uint32

File: internal/nodes/avi_model_nodes_utils.go [hunk 9]
```diff
@@ -28,28 +28,61 @@ import (
...
+	NetworkProfileRef             *string
+	NetworkSecurityPolicyRef      *string
+	OauthVsConfig                 *v1alpha2.OAuthVSConfig
+	PerformanceLimits             *v1alpha2.PerformanceLimits
+	RemoveListeningPortOnVsDown   *bool
+	SamlSpConfig                  *v1alpha2.SAMLSPConfig
+	SecurityPolicyRef             *string
 	Services                 []*v1alpha2.Service
-	SsoPolicyRef             *string
+	SslSessCacheAvgSize           *int32
```
[the remark this answers is not in the document]
Comment (arihantg) [reply internal/nodes/avi_model_nodes_utils.go]: updatd to *uint32

File: internal/nodes/avi_model_nodes_utils.go [hunk 10]
```diff
@@ -151,6 +202,30 @@ func (o *AviVsNodeCommonFields) ConvertToRef() {
 	}
 }
 
+func (o *AviVsNodeGeneratedFields) ConvertL7RuleFieldsToNil() {
```
[the remark this answers is not in the document]
Comment (arihantg) [reply internal/nodes/avi_model_nodes_utils.go:205]: These are not auto-generated.

File: pkg/apis/ako/v1alpha2/l7rule.go [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+
+type L7RuleSpec struct {
+	AllowInvalidClientCert        *bool              `json:"allowInvalidClientCert,omitempty"`
+	BotPolicyRef                  *string            `json:"botPolicyRef,omitempty"`
+	CloseClientConnOnConfigUpdate *bool              `json:"closeClientConnOnConfigUpdate,omitempty"`
+	HostNameXlate                 *string            `json:"hostNameXlate,omitempty"`
+	IgnPoolNetReach               *bool              `json:"ignPoolNetReach,omitempty"`
+	MinPoolsUp                    *int32             `json:"minPoolsUp,omitempty"`
```
[the remark this answers is not in the document]
Comment (arihantg) [reply pkg/apis/ako/v1alpha2/l7rule.go]: updated to *uint32

File: pkg/apis/ako/v1alpha2/l7rule.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+	PerformanceLimits             *PerformanceLimits `json:"performanceLimits,omitempty"`
+	RemoveListeningPortOnVsDown   *bool              `json:"removeListeningPortOnVsDown,omitempty"`
+	SecurityPolicyRef             *string            `json:"securityPolicyRef,omitempty"`
+	SslSessCacheAvgSize           *int32             `json:"sslSessCacheAvgSize,omitempty"`
```
[the remark this answers is not in the document]
Comment (arihantg) [reply pkg/apis/ako/v1alpha2/l7rule.go]: updated to *uint32

File: internal/nodes/avi_model_nodes_utils.go (hunk 2 above)
Comment (pkoshtavmware) [reply internal/nodes/avi_model_nodes_utils.go:209]: Problem wont be there in Uts, but might be seen during FTs, once you do end to end testing. 
Specifically during object creation at controller's end. I have faced issues in tha past when everything looked okay at AKO's end but still objects didnt get created at controller.
We can close this comment once you have tested CRUD end to end with these values.

**Review by DixitAakash — APPROVED** [review]

**Review by pkoshtavmware — APPROVED** [review]

**Comment by arihantg** [comment]
build ako

**Comment by arihantg** [comment]
build ako

**Review by DixitAakash — DISMISSED** [review]
LGTM

**Review by pkoshtavmware — DISMISSED** [review]
LGTM

File: internal/nodes/crd_translator.go [hunk 13]
```diff
@@ -479,3 +489,29 @@ func BuildL7SSORule(host, key string, vsNode AviVsEvhSniModel) {
...
+		utils.AviLog.Debugf("key: %s, msg: No L7Rule found for virtualhost: %s msg: %v", key, host, err)
+		deleteL7RuleCase = true
+	} else if l7Rule.Status.Status == lib.StatusRejected {
+		// do not apply a rejected L7Rule, this way the VS would retain
+		return
+	}
+	generatedFields := vsNode.GetGeneratedFields()
+	if !deleteL7RuleCase {
+		utils.AviLog.Debugf("key: %s, msg: applying l7 Rule %s", key, l7Rule.Name)
+		copier.CopyWithOption(vsNode, &l7Rule.Spec, copier.Option{DeepCopy: true})
```
Comment (akshayhavile) [inline internal/nodes/crd_translator.go:506]: This VSnode can be a sharedVS(Parent). With copy option, we are copying all fields to vsnode. How are we making sure that only fields applicable to parent VS is stored and rest is nullified?

File: internal/nodes/dequeue_ingestion.go [hunk 14]
```diff
@@ -40,6 +40,11 @@ func DequeueIngestion(key string, fullsync bool) {
 	sharedQueue := utils.SharedWorkQueue().GetQueueByName(utils.GraphLayer)
 
 	objType, namespace, name := lib.ExtractTypeNameNamespace(key)
+
+	if objType == lib.L7Rule {
```
Comment (akshayhavile) [inline internal/nodes/dequeue_ingestion.go:44]: This logic will be hit only from `FullSyncK8s`. As we are not doing anything here, just returning. Can we take out going through L7Rule logic from FullSyncK8s (Keep validation only don't enqueue the key).

Comment (akshayhavile) [reply internal/nodes/dequeue_ingestion.go:44]: Lets say L7Rule changed from valid to invalid or invalid to valid during AKO bootup. How are we handling that transition like applying L7Rule or taking out L7Rule from VS fields.

File: internal/objects/crdrules.go [hunk 15]
```diff
@@ -361,3 +365,31 @@ func (c *CRDLister) UpdateFQDNSSORuleMapping(fqdn string, ssoRule string) {
...
+	c.GetL7RuleToHostRuleMapping(l7Rule)
+}
+
+func (c *CRDLister) UpdateL7RuleToHostRuleMapping(l7Rule string, hostRule string) {
+	c.NSLock.Lock()
+	defer c.NSLock.Unlock()
+	_, hostRules := c.GetL7RuleToHostRuleMapping(l7Rule)
+	hostRules[hostRule] = true
+	c.L7RuleHostRuleCache.AddOrUpdate(l7Rule, hostRules)
+	c.GetL7RuleToHostRuleMapping(l7Rule)
```
Comment (akshayhavile) [inline internal/objects/crdrules.go:394]: Is this required?

File: internal/status/crd_status.go [hunk 16]
```diff
@@ -349,3 +349,74 @@ func SSORuleEventBroadcast(vsName string, vsCacheMetadataOld, vsMetadataNew lib.
...
+			return
+		}
+		UpdateL7RuleStatus(key, updatedL7RuleObj, updateStatus, retry+1)
+	}
+
+	utils.AviLog.Infof("key: %s, msg: Successfully updated the L7Rule %s status %+v", key, l7Rule.Name, utils.Stringify(updateStatus))
+}
+
+// L7RuleEventBroadcast is responsible from broadcasting L7Rule specific events when the VS Cache is Added/Updated/Deleted.
+func L7RuleEventBroadcast(vsName string, vsCacheMetadataOld, vsMetadataNew lib.CRDMetadata) {
```
Comment (akshayhavile) [inline internal/status/crd_status.go:386]: This seems to be not called from anywhere.

File: tests/evhtests/l7_evh_crd_test.go [hunk 17]
```diff
@@ -1869,3 +1869,445 @@ func TestGoodToBadSSORuleForEvh(t *testing.T) {
...
+	if err := lib.AKOControlConfig().V1alpha2CRDClientset().AkoV1alpha2().L7Rules("default").Delete(context.TODO(), l7ruleName, metav1.DeleteOptions{}); err != nil {
+		t.Fatalf("error in deleting l7Rule: %v", err)
+	}
+	if err := lib.AKOControlConfig().V1alpha2CRDClientset().AkoV1alpha2().L7Rules("default").Delete(context.TODO(), l7ruleName2, metav1.DeleteOptions{}); err != nil {
+		t.Fatalf("error in deleting l7Rule: %v", err)
+	}
+	TearDownIngressForCacheSyncCheck(t, modelName)
+}
+
+func TestValidToInvalidL7rule(t *testing.T) {
```
Comment (akshayhavile) [inline tests/evhtests/l7_evh_crd_test.go:2195]: Should we have test cases for 
1. PArent VS related properties
2. If hostrule deleted and L7Rule kept as it is, what is behaviour.

Comment (arihantg) [reply tests/evhtests/l7_evh_crd_test.go:2195]: added test case for dedicated vs and the 2nd point. will add for parent VS in a seperate Pr

File: internal/status/crd_status.go (hunk 16 above)
Comment (arihantg) [reply internal/status/crd_status.go:386]: Keeping it as discussed

File: internal/objects/crdrules.go (hunk 15 above)
Comment (arihantg) [reply internal/objects/crdrules.go]: removed this

File: internal/nodes/dequeue_ingestion.go (hunk 14 above)
Comment (arihantg) [reply internal/nodes/dequeue_ingestion.go:44]: removed the enqueue logic from FullSyncK8s. This transition will be handled when hostrule will be processed.
L7 rule is validated before hostrule so latest ststus of L7 rule will be used in hostrule processing.

File: internal/nodes/crd_translator.go [hunk 18]
```diff
@@ -223,6 +224,15 @@ func BuildL7HostRule(host, key string, vsNode AviVsEvhSniModel) {
 	serviceMetadataObj := vsNode.GetServiceMetadata()
 	serviceMetadataObj.CRDStatus = crdStatus
 	vsNode.SetServiceMetadata(serviceMetadataObj)
+
+	if lib.IsEvhEnabled() && !deleteCase {
+		generatedFields := vsNode.GetGeneratedFields()
```
[the remark this answers is not in the document]
Comment (arihantg) [reply internal/nodes/crd_translator.go]: moved this line in else block

File: internal/nodes/crd_translator.go [hunk 19]
```diff
@@ -223,6 +224,15 @@ func BuildL7HostRule(host, key string, vsNode AviVsEvhSniModel) {
 	serviceMetadataObj := vsNode.GetServiceMetadata()
 	serviceMetadataObj.CRDStatus = crdStatus
 	vsNode.SetServiceMetadata(serviceMetadataObj)
+
+	if lib.IsEvhEnabled() && !deleteCase {
```
[the remark this answers is not in the document]
Comment (arihantg) [reply internal/nodes/crd_translator.go]: moved

File: internal/nodes/avi_model_nodes_utils.go [hunk 20]
```diff
@@ -74,21 +111,32 @@ func (v *AviVsNodeGeneratedFields) CalculateCheckSumOfGeneratedCode() uint32 {
...
 
+	if v.SslSessCacheAvgSize != nil {
+		checksumStringSlice = append(checksumStringSlice, strconv.Itoa(int(*v.SslSessCacheAvgSize)))
+	}
+
 	if v.Services != nil {
 		checksumStringSlice = append(checksumStringSlice, utils.Stringify(v.Services))
 	}
-
+
```
[the remark this answers is not in the document]
Comment (arihantg) [reply internal/nodes/avi_model_nodes_utils.go]: removed

File: internal/nodes/crd_translator.go (hunk 13 above)
Comment (arihantg) [reply internal/nodes/crd_translator.go:506]: we are doing nullification for child vs properties which controller restricts . for Dedicated and parent vs all properties can be applied.

**Comment by arihantg** [comment]
build ako

**Comment by arihantg** [comment]
build ako

**Comment by arihantg** [comment]
build ako

**Comment by arihantg** [comment]
build ako

**Comment by arihantg** [comment]
build ako

**Comment by arihantg** [comment]
build ako

**Comment by arihantg** [comment]
build ako

**Review by akshayhavile — APPROVED** [review]
LGTM

**Review by DixitAakash — APPROVED** [review]
LGTM

**Pull request merged**